pkg/typesystem/typesystem.go

```diff
@@ -126,39 +126,37 @@ func Difference(base *openfgapb.Userset, sub *openfgapb.Userset) *openfgapb.User
 }
 
 type TypeSystem struct {
-	model           *openfgapb.AuthorizationModel
+	modelID         string
 	schemaVersion   string
 	typeDefinitions map[string]*openfgapb.TypeDefinition
-}
 
-type ValidatedTypeSystem struct {
-	*TypeSystem
+	// used only for validation
+	duplicateTypes []string
 }
 
 // New creates a *TypeSystem from an *openfgapb.AuthorizationModel.
 func New(model *openfgapb.AuthorizationModel) *TypeSystem {
+	duplicateTypes := make([]string, 0)
 	tds := map[string]*openfgapb.TypeDefinition{}
 	for _, td := range model.GetTypeDefinitions() {
+		if _, duplicateKey := tds[td.GetType()]; duplicateKey {
```

Comment:
Instead of carrying the duplicateTypes around and waiting for validation to fail, will it make more sense to just return error?

Reply:
that would change the signature of `New()` and i don't want to do that yet. But yeah this file deserves another refactor. The only reason we need this is for `Validate()` and this is only called in `WriteAuthorizationModel`

Reply:
Ok.

Reply:
Introduced a new constructor method that does validations